### PR TITLE
Prefix compiler-libs

### PR DESCRIPTION
**Pull request by nojb**
This PR proposes to prefix `compiler-libs`. 

### Why?

- To get rid of the `expunge` process: this is necessary to allow several interesting improvements to the toplevel experience (see [MPR#7589](https://caml.inria.fr/mantis/view.php?id=7589) and [MPR#6704](https://caml.inria.fr/mantis/view.php?id=6704))

- `compiler-libs` includes many modules with common names (`Path`, `Location`, `Parser`, etc). Prefixing it will be an improvement for all users.

### Proposal

As written above, we put all the modules in `compiler-libs` under a prefix. Actually we use four different top-level modules:

- `Ocaml_common` corresponding to `ocamlcommon.cma`
- `Ocaml_bytecomp` corresponding to `ocamlbytecomp.cma`
- `Ocaml_optcomp` corresponding to `ocamloptcomp.cma`
- `Ocaml_toplevel` corresponding to `ocamltoplevel.cma`

### Backwards compatibility

This will break all users of `compiler-libs`. To maintain backwards compatibility both the prefixed and unprefixed versions are distributed as a transitional measure. In any case, the unprefixed version is the one used to build the compiler (to avoid invasive changes to the build system).

### Approach

- The idea is **not** to touch the build of the compiler at all, so a parallel, prefixed, version of `compiler-libs` is built inside a special directory, `prefixed_compilerlibs`.

- The prefix modules are generated at `prefixed_compilerlibs/ocaml_{common,bytecomp,optcomp,toplevel}.ml` using the script `tools/gen_prefix.ml`.

- Each file, eg `bytecomp/lambda.ml` is copied to a prefixed source `prefixed_compilerlibs/ocaml_common__lambda.ml`, and then we use the official recipe for compiling prefixed libraries as explained in the manual. This way one can use `ocamldep` directly on the prefixed sources to compute dependendencies.

- The implementation uses `Makefile` macros to tame some of the duplication.

### Current state

The patch is complete:

- The prefixed compiler libs are built during `make world` and `make world.opt`
- Dependencies are computed with `make depend`
- The prefixed libraries are installed with `make install` in per-library directories under the `stdlib` install directory, as in #1569. For example, the prefixed `ocamlcommon` library files are installed in `+ocamlcommon`.

### Odds and ends

- If we keep the four top-level modules as above, it seems natural to also include an underscore in the archive/library names.
- Do we want to elevate `ocamlmiddleend` (currently used only for linking `ocamlobjinfo`) to a first-class member of `compiler-libs`?

### Related projects

- https://github.com/janestreet/ocaml-compiler-libs

@diml @dbuenzli

### CODE REVIEW

**Comment by nojb** [comment]
I pushed a simplified version of the PR (build the prefixed version in parallel to the unprefixed one). This way the existing build system is not touched at all which makes the change much easier to digest and we can trivially distribute both versions for backwards-compatibility. I have updated the PR description in consequence and removed the [POC] from the PR title since I am happy with this approach.

**Comment by nojb** [comment]
Taking a cue from #1569, I adapted the `install` target to install the prefixed `ocamlcommon` into `+ocamlcommon`, `ocamlbytecomp` into `+ocamlbytecomp`, etc. This way the prefixed and unprefixed version of `compiler-libs` can be installed simultaneously.

**Comment by nojb** [comment]
Some rough timings for `make world.opt` from a fresh checkout:

This PR:
```
real	7m37.383s
user	6m36.207s
sys	0m48.595s
```

trunk:
```
real	4m55.897s
user	4m12.304s
sys	0m34.882s
```

**Comment by ghost** [comment]
Thanks for working on this! I'm really looking forward not having to expunge toplevels as this leads to really confusing situations for users.

I wanted to give this a try in utop, which made me wonder how the new prefixed libraries should be presented to the user. Have you thought about this? The current prefixed libraries are available as `compiler-libs.ocamlcommon`,  ..., should we make the new ones available as just `ocaml_common`, `ocaml_bytecomp`, ...?

BTW, if we compile the two versions in parallel, then we won't be able to easily share values between code using the unprefixed version and code using the prefixed version. This is not a huge deal, but we could avoid it generating the following implementations for modules of the unprefixed version:

```
[@@@deprecated "use Ocaml_common.X instead"]
include Ocaml_common.X
```

This should also cut build times.

> If we keep the four top-level modules as above, it seems natural to also include an underscore in the archive/library names.

Maybe we should keep it as well for the sub-directory names?

**Comment by shindere** [comment]
Many thanks for this great piece of work, @nojb!

As you mentionned dependencies and just out of curiosity: according to
you, how much simpler would your task have been, were there no .depend files
to take into account?

**Comment by nojb** [comment]
Thanks for the comments!

> I wanted to give this a try in utop, which made me wonder how the new prefixed libraries should be presented to the user. Have you thought about this? The current prefixed libraries are available as `compiler-libs.ocamlcommon`, ..., should we make the new ones available as just `ocaml_common`, `ocaml_bytecomp`, ...?

Yes, that's one possibility. Another one could be `ocaml.common`, `ocaml.byteopt`, etc.

> BTW, if we compile the two versions in parallel, then we won't be able to easily share values between code using the unprefixed version and code using the prefixed version. This is not a huge deal, but we could avoid it generating the following implementations for modules of the unprefixed version:
> 
> ```
> [@@@deprecated "use Ocaml_common.X instead"]
> include Ocaml_common.X
> ```
> This should also cut build times.

This is a neat idea; I will give it a try.

> > If we keep the four top-level modules as above, it seems natural to also include an underscore in the archive/library names.
> 
> Maybe we should keep it as well for the sub-directory names?

Indeed.

**Comment by nojb** [comment]
> As you mentionned dependencies and just out of curiosity: according to
> you, how much simpler would your task have been, were there no .depend files
> to take into account?

Dealing with dependencies took about 50% of the time so far, I would say.

Do you have something particular in mind?

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2019/01/14 13:55 +0000):
> > As you mentionned dependencies and just out of curiosity: according to
> > you, how much simpler would your task have been, were there no .depend files
> > to take into account?
> 
> Dealing with dependencies took about 50% of the time so far, I would say.
> 
> Do you have something particular in mind?

Yeah, one of the thing that's waiting on my to-do-list for a very long
time is to try computing the dependencies dynamically rather than
storing them, but as I see it that's quite a lot of work so not sure it
will happen soon.

**Comment by nojb** [comment]
I pushed one more commit implementing most of @diml's suggestion.

#### Differences with previous commit:

- The prefixed `compiler-libs` are built directly into `compilerlibs` (instead of `prefixed_compilerlibs`) to signal that they are the "main" artifact.

- Unprefixed files are generated in `unprefixed_compilerlibs/XXX.{ml,mli}`, e.g. `unprefixed_compilerlibs/parsetree.mli` containing
  ```ocaml
  include module of struct include Ocaml_common.Parsetree end
  ```
- The unprefixed `compiler-libs` are built out of these generated modules so they have types compatible with the prefixed ones, and compilation is much faster.

- The variables `COMMON`, `BYTECOMP`, etc in the main `Makefile` now contain lists of `modules` (no extension), and include even those which are "interface-only". Two new variables `MLI_ONLY` and `ML_ONLY` are used to keep track those modules which are interface- or implementation-only.

- The compiler itself (and most associated tools) are built using the unprefixed `compiler-libs`. Note that the unprefixed `compiler-libs` are not yet marked as deprecated.

#### Still left to do adapt:

- [x] `make install`
- [x] `parsing/parser.ml` (due to `menhir --infer`)
- [x] `ocamlmiddleend`
- [x] `ocamlopttoplevel` (these libraries have modules overlapping with the other `compiler-libs`, which requires some adjusting)
- [x] `ocamldoc`
- [x] `debugger`
- [x] `manual` **[disable generation of doc for `compiler-libs` for now - needs more support from `ocamldoc`]**
- [x] `otherlibs/dynlink`
- [x] `testsuite`
- [ ] `make check_all_arches`

**Comment by dbuenzli** [comment]
> This will break all users of `compiler-libs`. To maintain backwards compatibility both the prefixed and unprefixed versions are distributed as a transitional measure.

How bad is that in practice ? Aren't most users using `compiler-libs` via the `ocamlfind` indirection ? If that is the case. Maybe only providing a new `META` file with the new names and tweaking the `META` file with the old names to point to the new ones with a deprecation warning predicate could be entirely sufficient no ?

**Comment by dbuenzli** [comment]
> * To get rid of the `expunge` process: this is necessary to allow several interesting improvements to the toplevel experience (see [MPR#7589](https://caml.inria.fr/mantis/view.php?id=7589) and [MPR#6704](https://caml.inria.fr/mantis/view.php?id=6704))

I'd to highlight that strictly speaking one does not need to get rid of the expunge process for improving the toplevel experience. Rather the proper functionality has to be exposed in the `Toploop` API and it should be made sure that this API does not leak expunged identifiers (more precisely exceptions) which is currently the case. I tend to think that it's maybe better to get rid of the expunge process but some people may think otherwise (e.g. depending on your viewpoint it puts more "garbage" in the initial environment).

**Comment by nojb** [comment]
> > This will break all users of `compiler-libs`. To maintain backwards compatibility both the prefixed and unprefixed versions are distributed as a transitional measure.
> 
> How bad is that in practice ? Aren't most users using `compiler-libs` via the `ocamlfind` indirection ? If that is the case. Maybe only providing a new `META` file with the new names and tweaking the `META` file with the old names to point to the new ones with a deprecation warning predicate could be entirely sufficient no ?

Even if you do something with the `META` files, note that you need to open some modules in your source code that you did not need to open before (e.g. `open Ocaml_common` to access module `Parsetree`). Existing code would be easy to adapt, but well...

I am also not overly worried about this point, but my feeling is that a good backwards-compatibility story will increase the chances of the PR being accepted. And in any case, @diml's [suggestion](https://github.com/ocaml/ocaml/pull/2218#issuecomment-454002724) seems a satisfactory answer to this question.

**Comment by dbuenzli** [comment]
> Even if you do something with the `META` files, note that you need to open some modules in your source code that you did not need to open before (e.g. `open Ocaml_common` to access module `Parsetree`).

Ah yes silly me.

**Comment by nojb** [comment]
> BTW, if we compile the two versions in parallel, then we won't be able to easily share values between code using the unprefixed version and code using the prefixed version. This is not a huge deal, but we could avoid it generating the following implementations for modules of the unprefixed version:
> 
> ```
> [@@@deprecated "use Ocaml_common.X instead"]
> include Ocaml_common.X
> ```
> This should also cut build times.

I implemented this. One small downside, though, is that those who link individual `.cmo`'s from the unprefixed lib will now have to pass the prefixed `.cmo` in addition. But I think it is an acceptable trade-off.

**Comment by nojb** [comment]
Scratch that, we don't install the individual `.cmo`'s, so there is no problem. It is only the `testsuite` that dares to do such things 😄

**Comment by nojb** [comment]
This is ready for review!

**Comment by Octachron** [comment]
@nojb , do you want me to have a look at the documentation part?

**Comment by nojb** [comment]
> @nojb , do you want me to have a look at the documentation part?

Yes, please!

The objective is that `ocamldoc` be able to generate documentation for several top-level prefixed libraries at once: `Stdlib`, `Ocaml_common`, `Ocaml_bytecomp`, etc. The modules `M__N` should appear only as `N` in the module index (maybe under different headings corresponding to the different top-level prefixed libraries) and the module `M__N` should be displayed as `M.N`, etc.

Thanks for the help!

**Comment by nojb** [comment]
@alainfrisch observed that if one distributes both an "unprefixed" and a "prefixed" version of `compiler-libs` (where the unprefixed modules just `include` the prefixed ones) then it is not possible to link libraries using these different versions since it would result in the prefixed modules linked twice.

Taking this into account, I will merge the two versions into a single library that contains both the prefixed and unprefixed modules, with the unprefixed modules deprecated. This will also simplify some things since only one library will have to be built and installed.

**Comment by hcarty** [comment]
If the prefixed and unprefixed versions all live in the same library, is there much benefit to the prefixed versions?  If I understand correctly, the unprefixed module names would still have name conflicts with other linked modules.

**Comment by nojb** [comment]
> If the prefixed and unprefixed versions all live in the same library, is there much benefit to the prefixed versions? If I understand correctly, the unprefixed module names would still have name conflicts with other linked modules.

You are right; the unprefixed modules are there only for backwards-compatibility (you can use them but you will get a deprecation alert). In the following version we would remove them and only the prefixed modules would be left. This is to avoid breaking every single use of `compiler-libs` out there.

If we were to decide we don't care we can skip the intermediate step and distribute only the prefixed modules from the get-go.

**Comment by hcarty** [comment]
Makes sense.  If I understand correctly:

* 4.08.x will have both prefixed + unprefixed versions bundled into one library.  Uses of the unprefixed modules will work but result in a deprecation warning.
* 4.09.x will have only the unprefixed version.  Uses of the unprefixed modules won't compile.

**Comment by nojb** [comment]
> Makes sense. If I understand correctly:
> 
> * 4.08.x will have both prefixed + unprefixed versions bundled into one library.  Uses of the unprefixed modules will work but result in a deprecation warning.
> * 4.09.x will have only the unprefixed version.  Uses of the unprefixed modules won't compile.

That's the plan (except that with 4.09 and 4.10 -- it is too late for inclusion in 4.08).

**Comment by hcarty** [comment]
Ah, of course re: 4.09/4.10.  Thanks!

**Comment by ghost** [comment]
I ran the following script at the base and tip of this feature:

```
$ git clean -dfx
$ git checkout $COMMIT
$ ./configure --prefix ~/ocamls/$COMMIT
$ make world.opt -j16
$ make install
$ cd ~/ocamls/$COMMIT
$ find . > /tmp/$COMMIT
```

and then I diffed the resulting files. As expected, the only difference is the addition of various `lib/ocaml/ocamlxxx` directories containing the prefixed object files. I also noticed the following files:

```
./lib/ocaml/ocamlbytecomp/main.o
./lib/ocaml/ocamlbytecomp/main.cmx
./lib/ocaml/ocamloptcomp/optmain.o
./lib/ocaml/ocamloptcomp/optmain.cmo
./lib/ocaml/ocamloptcomp/optmain.cmx
```

@nojb are these expected?

**Comment by nojb** [comment]
> @nojb are these expected?

No, they are not. Probably a glitch in the `install` rule. I will fix this (after rebasing - lots of conflicts with `trunk`).

**Comment by nojb** [comment]
By the way, did you run your script with the current version of the PR? I don't think you should be getting those directories in the latest version; in fact there should be no diff at all in the install dir between trunk and this PR (since prefixed and unprefixed modules are put in the same library).

**Comment by ghost** [comment]
I used commit 711362988 as base and 119335aa9 as tip. It's possible I forgot to upgrade my local branch...

**Comment by ghost** [comment]
I re-ran the script with the new revisions. I noticed that the cmt files for the unprefixed modules are no longer installed. Is that intentional? I would suggest to get rid of them at the same time as the unprefixed modules.

I also noticed that all the man pages for the compiler modules are gone. Personally, I'm fine with this change, but it might be worth mentioning it in the changelog.

**Comment by nojb** [comment]
> I re-ran the script with the new revisions. I noticed that the cmt files for the unprefixed modules are no longer installed. Is that intentional? I would suggest to get rid of them at the same time as the unprefixed modules.

Thanks for the review. I pushed a fix for this along with several simplifications.

> I also noticed that all the man pages for the compiler modules are gone. Personally, I'm fine with this change, but it might be worth mentioning it in the changelog.

Yes, they are disabled because `ocamldoc` cannot cope very well with prefixed libraries. @Octachron said he may have a look into adapting `ocamldoc`, but I would also not be shocked if we simply do not include docs for `compiler-libs` in the manual.

**Comment by garrigue** [comment]
This may not be directly related to this PR, but at least `ocamlcommon.cma` could use plain `-pack` rather than this approach: it is linked with `-linkall`.
Is there anywhere an analysis of why this `-linkall` is needed?
It dramatically grows the size of executables using it, and prevents from linking with `ocamlcommon.cma` in the `tools` directory.

**Comment by nojb** [comment]
> This may not be directly related to this PR, but at least `ocamlcommon.cma` could use plain `-pack` rather than this approach: it is linked with `-linkall`.
> Is there anywhere an analysis of why this `-linkall` is needed?
> It dramatically grows the size of executables using it, and prevents from linking with `ocamlcommon.cma` in the `tools` directory.

Good question, I am not sure; will try to do some git archaeology to see if it reveals anything.

**Comment by nojb** [comment]
> This may not be directly related to this PR, but at least `ocamlcommon.cma` could use plain `-pack` rather than this approach: it is linked with `-linkall`.
> Is there anywhere an analysis of why this `-linkall` is needed?
> It dramatically grows the size of executables using it, and prevents from linking with `ocamlcommon.cma` in the `tools` directory.

The `-linkall` flag was added in https://github.com/ocaml/ocaml/pull/53

**Comment by xavierleroy** [comment]
Recently (more or less) we added support for `-linkall` on individual compilation units (.cmo and .cmx files), not just on library files (.cma and .cmxa).  So it would be possible to set `-linkall` on the compiler units that really need it because they self-initialize, instead of on the whole `ocamlcommon.cm{,x}a` library.  Whether this would save significant code size is unclear, however.

File: Makefile [hunk 1]
```diff
@@ -723,19 +511,93 @@ manual-pregen: opt.opt
 # The clean target
 clean:: partialclean
 
+# Prefixed compiler-libs
+
+compilerlibs/ocaml_common__compdynlink.cmo: \
+	compilerlibs/ocaml_common__compdynlink.mlbyte
+	$(CAMLC) $(COMPFLAGS) $(shell compilerlibs/Compflags $@) -c -impl $<
```
Comment (damiendoligez) [inline Makefile:518]: @shindere will be sad to see a new `Compflags` script added to the build system...

**Comment by damiendoligez** [comment]
Please check this on MacOS before you merge (for example with precheck) because Apple's version of gnumake is rather old and the Makefile changes are non-trivial.

**Comment by shindere** [comment]
Damien Doligez (2019/04/10 10:17 +0000):
> damiendoligez commented on this pull request.
> 
> 
> 
> > @@ -723,19 +511,93 @@ manual-pregen: opt.opt
>  # The clean target
>  clean:: partialclean
>  
> +# Prefixed compiler-libs
> +
> +compilerlibs/ocaml_common__compdynlink.cmo: \
> +	compilerlibs/ocaml_common__compdynlink.mlbyte
> +	$(CAMLC) $(COMPFLAGS) $(shell compilerlibs/Compflags $@) -c -impl $<
> 
> @shindere will be sad to see a new `Compflags` script added to the
> build system...

Indeed. Well initially it was @xavierleroy's suggestion to get rid of
such scripts, see #8601.

**Comment by gasche** [comment]
I had sort of forgotten about this PR. For the record, I have been bitten by compiler-units-name conflicts related to compilerlibs recently (I had a `lexer.mll` file; the fix was to tell Dune to use `(wrapped_executables true)`, so that the project's units were prefixed, avoiding the conflict), and I use JaneStreet's [ocaml-compiler-libs](https://opam.ocaml.org/packages/ocaml-compiler-libs/) package in outside projects using compiler-libs, so I already use say `Ocaml_common.Lexer`; this coding style is compatible with prefixing+aliases, and even if we don't do actually use prefixing in the compilerlibs today, I think we should make it possible for downstream users -- or clearly advertise ocaml-compiler-libs as the right middle-man.

From a high-level perspective, I think it's a royal pain that all OCaml projects have to consider unit prefixing/wrapping; Dune makes it somewhat hidden, but they remain ugly workarounds for the lack of a proper namespacing system -- or at least support mixing compilation units with the same file name and different digests in the same project (my own #2228 was a zero-th step in this direction, and I know @mshinwell and @lpw25 are working on this on-and-off as well).

**Comment by nojb** [comment]
I closed this PR because it was seriously out of date, but I am planning to take another shot at this shortly, following your #9211 which has rearranged some things in a useful way.

**Pull request closed without merging**